### ISSUE TITLE
harvesters / tractors not turning correctly in field work mode

### ISSUE THREAD

**Issue by Opa-Andre** [Courseplay/courseplay#3434]
I've observed an issue in field work mode, which is always reproducable but I don't know the reason for.
I just reproed it again using the following environment, but it happens as well when using other harvesters or tractors.

FS19  Version 1.2 (Steam)
CP Version 6.01.00139
Map:  Felsbrunn
4 x Fendt 1000 tractors equipped with Agrisem Combiplow Gold 4m (Tiefenlockerer).
No mods installed / used during repro

Steps to Repro: 
Drive all equipped tractors to NW edge of field 12 and generate the course as follows:
Field work course for 4 vehicles, starting from actual waypoint (NW of Field 12), no headlands - so the generated course just goes from West to East, turn to South, next lane from East to West, turn to South, drive lane from West to East and so forth.

Load the course into all 4 tractors and activate symmetric lane change. 
First tractor should start uses lane right 2, second tractor lane right 1, third tractor lane left 1 and fourth tractor lane left 2. I also activated Convoy mode and "Turn on Field" mode, but the issue is independent of these settings.

The issue you will observe occurs only using the tractor number 2 which starts in right lane1.
This tractor will "forget" to change its lane turn **only** on the East turn of the field, so after the first turn he will still use "lane 1 right" instead of changing it to "lane 1 left". So after this turn he will use the same lane as tractor #3. All other tractors are using their correct turn settings (tractor #1, #3 and #4). If you manually stop tractor 2 on the east side of the field, set it up to use the correct "lane 1 left", and restart him using the next possible waypoint, he will drive the correct lane to the West side of the field, turn correctly back to "lane 1 right" drive the lane from West to East correctly as well, but then on the East side he will again forget to change the lanes. You will also note that that tractor will turn nearly 1 meter before the end of the lane and thus probably not getting his "change lane" command.

As I said before this issue can be reproduced also using other equipment - So the same issue happens for example when using more than one "Ropa Panther" harvesters to harvest sugar beet on that field which are also from the base game (no mods).

If you need more details, please let me know.

**Comment by sootysax** [comment]
@Opa-Andre Do you think this is similar to my issue #3382 ?

**Comment by Opa-Andre** [comment]
I'm unsure - it sounds similar but as I don't know how many harvesters you were using I cannot confirm for sure. IIRC in my example it always happened only to the tractor with the closest right offset to the "middle" of the projected lanes. So when generating a course for 4 vehicles, it happens only to vehicle number 2... when using 2 vehicles, it happens to vehicle 1.  As I'm not familiar with the debug levels I tried to explain all details with default equipment, so at least it should be easy to reproduce for a developer. If you want me to set it up using specific debug levels and to provide logfiles, just let me know and I'll do.

**Comment by Messy71** [comment]
@Opa-Andre it would be nice for the devs if you’ve created a savegame with 4 different tractors ( to be distinguished easily in the log file) and align them up right before the edge  where the problem happens (by this i mean with courses loaded, set to start from nearest waypoint [ align them 7/8 waypoints before the turn waypoint] and specify the order in wich you start them[ outside first or inside first])

**Comment by pvaiko** [comment]
@Messy71 oh yes, would make life so much eaisier...

**Comment by Opa-Andre** [comment]
@pvaiko will do... just came home :)

**Comment by Opa-Andre** [comment]
@pvaiko Here you are. I've prepared the scenario for easy repro. The zip file contains:
- log.txt
- savegame1 directory including files within
- mods directory (with the FS19_Courseplay.zip file)
- Courseplay_Courses directory with the needed files to repro

To repro I've used 4 different tractors with the same equipment (front weight in different colors and Tiefenlockerer).
When you "TAB" into the first tractor with the RED front weight, you'll need to load the course for it as it somehow forgot the course during save. All other settings are set correctly. Start the tractor to drive its course. Tab into the second tractor (YELLOW front weight), he should already correctly set) and start its course. Tab into the third tractor (GREEN front weight) and start its course. Finally tab into the fourth tractor and start its course. 

You'll note the bug occurring with the second (YELLOW front weight) tractor only at the next turn. It will start its turn a bit early and won't change its lane as defined. (From 1 right to 1 left). All other tractors are working as expected. Now jump into this tractor and change it's lane manually from "1 right" to "1 left". He will follow the course and at the next turn he will correctly change the lane from "1 left" back to "1 right" ... but after this correct turn on the West side of the field, he will run into the same bug again at at the East side turn.

Unfortunately I wasn't able to create a gist with the repro file, so I've saved it in my public OneDrive folder. Here the link to the [Courseplay_repro.zip](https://1drv.ms/u/s!ApAS1kHbd61LgbIiKDSLycOgBqavBw) for downloading it.

If you need more details etc. please let me know.

**Comment by pvaiko** [comment]
@Opa-Andre great setup, thank you, unfortunately I can't reproduce it. Everything works beautifully when I start them, so much fun to watch with the convoy option turned on...

I noticed however that after loading the game, the yellow one had 2 Right lane offset, the green 1 Right and the blue 1 Left, no idea why but I corrected these before starting them and everything worked perfectly.

**Comment by Opa-Andre** [comment]
Hi @pvaiko ,

strange that you couldn't repro it. Regarding the setup with the wrong lane offsets, I can only imagine you did reload the course for the tractors - > this often leads to lane changes. I just checked the vehicles.xml in the repro files and there the offsets are set correctly. Anyway this shoudln't matter - so I'm wondering, what's different... 

For your convenience I just reproed it again and did a screen recording. So you can see what happens on my system. I had to scale down the recording as the original was in UltraHD and 4GB in size, but you should notice what I mean. I just started the game - rechecked the settings and started the tractors by tapping through them. You'll notice that the second one (yellow) just stops before the end of the lane and turns it wheels, then waits for something before continuing without lane change. Therefore after the turn it will use the same lane as the Green one, which is turning correctly.
Please wait a few seconds when clicking the [repro movie link](https://1drv.ms/v/s!AJAS1kHbd61LgbIk) as it will somehow initialize the web player first..during that time you'll have a white screen. Alternatively you can also use the folder link from the repro files I posted in the last log - there you'll have direct access to the video clip as well.

kind regards
Andre

**Comment by pvaiko** [comment]
Weird. Only thing we can do is you turn on debug channels 12 and 14 and post the log here, best would be a video as well as we'd see what that yellow guy is doing instead of a proper turn maneuver.

**Comment by Opa-Andre** [comment]
Agree, weird. For which vehicles should I turn the debug levels on - the yellow one only, #1 and #2, #2 and #3 or all of them?

**Comment by pvaiko** [comment]
Yellow one only.

**Comment by Opa-Andre** [comment]
Here they are. 
[Repro3 Video](https://1drv.ms/v/s!AJAS1kHbd61LgbIm)
[Repro3 Logfile](https://1drv.ms/u/s!AJAS1kHbd61LgbIl)

I used the original repro files as before, but the actual last version of CP (6.01.00144).
During repro I noticed that switching on the debug levels on the yellow one seems to be a global setting due to the results, but you'll see this in the movie clip.

**Comment by pvaiko** [comment]
Perfect, thank you. As with most weird software problems this is a timing issue, the convoy mode interferes with the turn, we don't handle that correctly, happens only when your timing is right when starting the vehicles.

**Comment by pvaiko** [comment]
If you want, you could try setting a speed field 10 or 9 for the yellow one and see what happens and post the log. Video is not needed. Thanks.

**Comment by Opa-Andre** [comment]
Well, seems I'm having a good (or bad) timing then on field 12. :)
Here the [log (same scenario) on field 10](https://1drv.ms/u/s!AJAS1kHbd61LgbIl) - recorded the right turns there as well. There I cannot reproduce the issue. All 4 tractors are turning as expected.

Especially weird (from my own observations) is, that I rarely can reproduce the issue. I've seen it on other fields rarely but I have to admit, that due to the mostly smaller field sizes I didn't use several fieldwork vehicles for the same work on one field. Only on field12 it is always reproducible for me. And there it's occurring always but only with the tractor or harvester which is running on lane 1 right.

**Comment by pvaiko** [comment]
The good news is that I now know why is this happening, it is due to the offset and the way the Pure Pursuit Controller works, as the turn start waypoints are pointing towards the turn end waypoints, they are offset towards the previous waypoint of the course and depending on a multitude of factors but especially the work width, this will lead to, at least in this case, the PPC skipping the turn start waypoint.

The bad news is that at the moment I have no idea how to fix this in an elegant way. And yes, it is the same as some of the issues in #3382 (although not the one in the title of #3382, that's an entirely other problem)

**Comment by MoslessNeo** [comment]
I have a similar problem, I always liked using giant fields for the job. 64 to 300 hectares.
In version 17 he used up to 8 workers in rectangular fields.
In FS19, the maximum that I get is 4 workers, being the ideal 3.
More than that, the workers 5, 6 ... get lost on the route and go to tangent.

**Comment by Opa-Andre** [comment]
Thanks much for your update, Peter! In regards of not finding an elegant way to fix it, I'd recommend to  put this bug on hold for a week or two, set a reminder, not to think about it during that time and do something else. Thats what I also do when not being able to find a viable solution.

**Comment by Opa-Andre** [comment]
Hi @pvaiko, as you've closed this issue, I used my repro scenario again to check. This time with CP version 6.01.00194. I can confirm happily, that the issue doesn't occur anymore - so it's really fixed!

Thanks much for solving this tricky timing issue!

Happy Easter
- Andre
